transformer_engine/pytorch/graph.py

```diff
@@ -436,19 +446,26 @@ def hook_fn(module, inputs, outputs):  # pylint: disable=unused-argument
                     per_callable_static_grad_outputs[per_callable_bwd_idx] = static_grad_outputs
                     per_callable_static_grad_inputs[per_callable_bwd_idx] = static_grad_inputs
 
-                    if io_memory_reduction:
-                        # Weak ref the static outputs and static grad inputs.
+                    # Weak ref the static outputs and static grad inputs that are no longer needed
+                    # in the following steps. These two type of tensors are both in cudagraph
+                    # mempool, so we just deallocate them and let PyTorch's memory allocator
+                    # reuse them elsewhere.
+                    if _reuse_graph_input_output_buffers:
+                        # Weak ref the static outputs of the forward pass of this backward. It's
+                        # no longer needed after the corresponding backward graph is built up.
                         per_callable_static_outputs[per_callable_bwd_idx] = make_weak_ref(
                             static_outputs
                         )
+                        # Weak ref the static grad inputs of the previous backward pass. The static
+                        # grad inputs should be kept alive for one more backward pass because its
+                        # value is still needed as the input of its next layer backward.
```

Comment:
Based on the discussion at https://github.com/NVIDIA/TransformerEngine/pull/1234#discussion_r2170504859:
```suggestion
                        # Weak ref the static grad inputs of the previous backward pass. 
                        # Note: After a backward pass, we assume Mcore will send the
                        # grad input to another pipeline parallel rank and that the 
                        # communication is finished before the end of the next backward
                        # pass.
```